Enable polling in file watcher

daniel-zullo-frequenz opened the pull request:
`inotify` does not work reliably with network file systems (e.g., NFS, CIFS) commonly used in cloud environments. These
file systems may not propagate file system events correctly, causing `inotify` to miss changes. To ensure consistent file
monitoring across these environments, polling is enabled by default in `FileWatcher`.

Fixes #256

daniel-zullo-frequenz committed:
Enable polling in file watcher

`inotify` does not work reliably with network file systems
(e.g., NFS, CIFS) commonly used in cloud environments. These
file systems may not propagate file system events correctly,
causing `inotify` to miss changes. To ensure consistent file
monitoring across these environments, polling is enabled by
default in FileWatcher.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

llucax commented on src/frequenz/channels/file_watcher.py:
```diff
@@ -127,13 +128,19 @@ def __init__(
         paths: list[pathlib.Path | str],
         event_types: abc.Iterable[EventType] = frozenset(EventType),
+        *,
+        enable_polling: bool = True,
```
If you are using a positive term then I would use the same as `watchfiles` (`force_polling`), as probably even when passing `force_polling=False` (`enable_polling=False`) polling will probably be used if I notify or other event based system calls are not available in the OS.

llucax commented on tests/test_file_watcher_integration.py:
Why disabling polling in tests? Are there any tests using polling?

daniel-zullo-frequenz replied:
The unit test is using polling. I can try to make one integration tests work at least after introducing the `polling_interval` parameter.

daniel-zullo-frequenz commented:
Updated based on initial suggestions from Luca

daniel-zullo-frequenz committed:
Remove redundant TARGETPLATFORM in Dockerfile

Setting platform to predefined $TARGETPLATFORM is
redundant as this is the default behavior.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
Added a new commit to remove the redundant `TARGETPLATFORM` in Dockerfile

llucax commented on .github/containers/test-installation/Dockerfile:
```diff
@@ -3,7 +3,7 @@
 # multiple platforms in the CI. It is not used to build the package itself.
 
-FROM --platform=${TARGETPLATFORM} python:3.11-slim
+FROM python:3.11-slim
```
:heart:

llucax reviewed (commented):
I have the same concern as with #321 , this is actually mixing one *bug fix* (setting `force_poll=True`, which depending on how we see it it could even be a breaking change[^1]) and a new feature (being able to configure `force_polling`).

[^1]: Maybe this is worth documenting in the `Upgrading` section, mentioning the default mechanism was changed and updates might have a bit of a delay now and watching a large amount of files might be less efficient, and letting the user know how to go back to the previous default.

llucax commented:
For simplicity I think we can release all of this in v1.2.0 and that's it.

daniel-zullo-frequenz commented:
Updated release-notes to add an entry in `Upgrading`. And rebased onto the latest v1.x.x

llucax commented on RELEASE_NOTES.md:
```diff
@@ -1,11 +1,19 @@
 
+## Upgrading
+
+- `FileWatcher`: The file polling mechanism is now enabled by default. This provides reliable and consistent file monitoring on network file systems (e.g., CIFS). However, it may have a performance impact on local file systems or when monitoring a large number of files.
```
```suggestion
- `FileWatcher`: The file polling mechanism is now forced by default. This provides reliable and consistent file monitoring on network file systems (e.g., CIFS). However, it may have a performance impact on local file systems or when monitoring a large number of files.
```

llucax commented on tests/test_file_watcher_integration.py:
```diff
@@ -113,7 +117,11 @@ async def test_file_watcher_exit_iterator(tmp_path: pathlib.Path) -> None:
+    file_watcher = FileWatcher(
+        paths=[str(tmp_path)],
+        force_polling=True,
+        polling_interval=timedelta(seconds=0.05),
```
50ms sounds a bit extreme, but I guess it's fine...

daniel-zullo-frequenz replied:
I know, the timer of the test is set to 100ms  so it needed a safe value to ensure the changes are picked. We can increase the timer interval but that'd make the test to take longer
